src/flag_gems/ops/cummin.py

```diff
@@ -0,0 +1,334 @@
+import logging
+import math
+
+import torch
+import triton
+import triton.language as tl
+import triton.language.core as core
+
+from ..utils import libentry
+from ..utils import triton_lang_extension as tle
+
+
+@triton.jit
+def tl_cummin(input, index, axis=0):
+    return core.associative_scan(
+        (input, index), axis, tle.minimum_with_index_tie_break_right
+    )
+
+
+@triton.jit
+def _argmin_combine_tie_break_right(value1, index1, value2, index2):
+    tie = value1 == value2 and index1 > index2
+    lt = value1 < value2 or tie
+    value_ret = core.where(lt, value1, value2)
+    index_ret = core.where(lt, index1, index2)
+    return value_ret, index_ret
+
+
+@triton.jit
+def tl_min_tie_break_right(input, axis=None, keep_dims=False):
+    return core._reduce_with_indices(
+        input, axis, _argmin_combine_tie_break_right, keep_dims=keep_dims
+    )
+
+
+@libentry()
+@triton.jit(do_not_specialize=["n_elements"])
+def add_base_min_kernel(
+    out,
+    out_indices,
+    partial_min,
+    partial_min_indices,
+    n_elements,
+    BLOCK_SIZE: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    offset = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    mask = offset < n_elements
+
+    out_ptrs = out + offset
+    out_indices_ptrs = out_indices + offset
+    out_vals = tl.load(out_ptrs, mask=mask)
+    out_indices = tl.load(out_indices_ptrs, mask=mask)
+
+    if pid > 0:
+        partial_min_ptrs = partial_min + pid - 1
+        last_part_min_via_min = tl.load(partial_min_ptrs)
+        partial_min_indices_ptrs = partial_min_indices + pid - 1
+        last_part_min_index_via_min = tl.load(partial_min_indices_ptrs)
+
+        final_vals = tl.minimum(out_vals, last_part_min_via_min)
+        final_indices = tl.where(
+            out_vals <= last_part_min_via_min, out_indices, last_part_min_index_via_min
+        )
+        tl.store(out_ptrs, final_vals.to(out_vals.dtype), mask=mask)
+        tl.store(out_indices_ptrs, final_indices, mask=mask)
+
+
+@libentry()
+@triton.jit(do_not_specialize=["n_elements"])
+def scan_part_min_kernel(
+    inp,
+    out,
+    out_indices,
+    partial_min,
+    partial_min_indices,
+    n_elements,
+    BLOCK_SIZE: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    offset = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    mask = offset < n_elements
+
+    inp_ptrs = inp + offset
+    inp_vals = tl.load(inp_ptrs, mask=mask, other=float("inf"))
+    if (
+        tl.constexpr(inp_vals.dtype.is_int64())
+        or tl.constexpr(inp_vals.dtype.is_uint64())
+    ) or tl.constexpr(inp_vals.dtype.is_fp64()):
+        inp_vals = inp_vals
+    elif tl.constexpr(inp_vals.dtype.is_int()):
+        inp_vals = inp_vals.to(tl.int32)
+    else:
+        inp_vals = inp_vals.to(tl.float32)
+    result, cummin_indices = tl_cummin(inp_vals, offset, axis=0)
```

Comment:
I have an idea to make the code easier to read and potentially improves the performance.

Let's take `M=N=1` as an example, the digram of kernels is 


There `scan_part_min_kernel` maybe called `N` times and util ther is only one block in the grid. Then call `scan_part_min_kernel` `N-1` times in a reverse order(the recusrsion here).

After each call to scan_part_min_kernel, you get an output(cummin values  & indices）whose size has been reduced by a factor of `1024`). But the indices are the indices at the original array.

Since you have used `reducetion` and `scan` with a combiner that takes care of the index, it may not be a arange from 0 to some extent. It is like a kind of argmin.

So I suggest not using `offset` in `tl_cummin(inp_vals, offset, axis=0)`. Instead load the input indices and `tl_cummin(inp_vals, input_indices, axis=0)`. Since after each call to this kernel, the input is shrinked, and the indices are not contiguous.

Loading the input indices maks this more staright-forward, without needing to reconstruct `true_indices` beblow. I believe that the reason why you need to reconstruct true_out_indices is that the `indices` used in `tl_cummin`ares not the indices in the original indices, instead, there are contiguous `offsets` in the new kernel. 

In a consequence, you have to load from `out_indices` (which is used also as input indices when you load from them). But as the `cummin_indices` is the result of `tl_cummin`, indirect memory access is needed, and the resulting addresses to load from are not single strided view.